src/Components/ResourcePicker/ResourcePicker.tsx

```diff
@@ -0,0 +1,649 @@
+                            ? 'https://' + resource.properties.hostName
+                            : null;
+                    case AzureResourceTypes.StorageAccount:
+                        return resource.properties.primaryEndpoints?.blob;
+                    default:
+                        return resource[
+                            AzureResourceDisplayFields[displayField] // storage blob containers does not have url information in its resource properties, it is constructed based on the storage account blob url
+                        ];
+                }
+            } else {
+                return resource[AzureResourceDisplayFields[displayField]];
+            }
+        },
+        [displayField]
+    );
+
+    // after fetching the resources merge them with existing options in the dropdown
+    const mergeOptionsWithProps = useCallback(
+        (options, additionalOptions, selectedOption) => {
+            if (additionalOptions) {
+                // do the merging with existing options: add the additional options manually entered by user to the end if it is not in the fetched data
+                const existingOptionTexts =
+                    resourcesRef.current?.map((resource) =>
+                        getDisplayFieldValue(resource)
+                    ) || [];
+
+                const optionsToAdd = additionalOptions.filter(
+                    (additionalOption) =>
+                        existingOptionTexts.findIndex((existingOptionText) =>
+                            isValuesEqual(
+                                existingOptionText,
+                                additionalOption.text
+                            )
+                        ) === -1
+                );
+
+                if (optionsToAdd.length) {
+                    if (resourcesRef.current?.length > 0) {
+                        // add the freeform options header if the resources are fetched and exists
+                        options.push(freeformOptionsHeader);
+                    }
+                    options = options.concat(optionsToAdd);
+                }
+            }
+            if (selectedOption) {
+                const selectedOptionInNewOptions = options.find((option) =>
+                    isValuesEqual(option.text, selectedOption.text)
+                );
+                if (!selectedOptionInNewOptions) {
+                    const freeFromOptionsHeaderExist = options?.find(
+                        (option) =>
+                            option.itemType ===
+                                SelectableOptionMenuItemType.Header &&
+                            option.text === freeformOptionsHeaderText
+                    );
+                    if (!freeFromOptionsHeaderExist) {
+                        options.push(freeformOptionsHeader);
+                    }
+                    options = options.concat([selectedOption]);
+                } else {
+                    setSelectedKey(
```

Comment:
it is confusing to me that in a function called "mergeOptionsWithProps" that we are setting the selected key. Why do we have to do that? Why would the key change when generating the list of items? Should't it match between the old and new list of options?

Reply:
`mergeOptionsWithProps` is called after fetching the actual data. it is setting the dropdown options after fetch finishes in this order => fetched resources + additional options using freeform + selected option (if exists and not included in the additional options). The reason why I set the selected key again is because initially the "key" is based on the display value since there is no "id" information of those dropdown options (read from localstorage or passed via props). So I need to update the selected key value with resource ids after fetching is completed and merging options is done, since there is no longer that option with the old key exist.